### DISCUSSION TITLE
Implementation of the gorqlite client in addition to database/sql

### DISCUSSION THREAD

**Discussion by maxnrb** [aarondl/sqlboiler#1223] [Q&A]
Hello, 
I'm trying to develop a driver for [rqlite database](https://github.com/rqlite/rqlite) with the [gorqlite client](https://github.com/rqlite/gorqlite). I succeeded in generating the sources

Now, I need to make some modifications to the sqlboiler project in order to support the gorqlite client in addition to database/sql. Indeed, rqlite doesn't support transactions for example (cf https://github.com/volatiletech/sqlboiler/blob/master/boil/db.go). Moreover, the return type "sql.Rows" is different with gorqlite.

I thought about several implementations: 
- create a "copy" of the "boil" package with different Executor interfaces, but not easy to maintain in the future
- create interfaces of the Executor interfaces with less information (only the Query() function) and which return an interface instead of "sql.Rows", like that:
```go
type BaseExecutor interface {
	Query(query string, args ...interface{}) (interface{}, error)
}

// Executor can perform SQL queries.
type Executor interface {
	BaseExecutor

	Exec(query string, args ...interface{}) (sql.Result, error)
	QueryRow(query string, args ...interface{}) *sql.Row
}
```

However, this includes modifying the generation of files by sqlboil, to replace the Executor interface by BaseExecutor, and it will be necessary for everyone to re-generate their source.

So I'm looking to make the necessary changes without creating any breaking change. If anyone has advice or help, I'm interested.

**Comment by stephenafamo** [comment]
The bigger issue with modifying the `Executor` interface is that `*sql.DB` will no longer implement it and you'd need a wrapper to pass to it every time. For this reason alone, I don't think the `Executor` interface will be changed.

There's no easy solution to use a client that doesn't follow the standard library. A popular example is [pgx](https://github.com/jackc/pgx), thankfully, pgx has the `/stdlib` sub-package to work with `database/sql`.

One possibility is to overwrite the templates in your driver, but that may be quite tedious to maintain.

Another possibility is to create a stdlib wrapper driver around the gorqlite client. I think this is the `cleanest` solution and may even be beneficial when working with any other package that plays nicely with `*sql.DB`.

**Comment by maxnrb** [comment]
I have started to develop a stdlib wrapper driver, agreeing with you that this is the best solution.

However, due to the design of rqlite, it is not really possible. For example, rqlite does not support transitions, it adds queues... ([see here for more details](https://github.com/rqlite/gorqlite#why-not-a-databasesql-driver)). So it's not really compatible with the sql stdlib implementation. Sadly, I think I will use another RDBMS.

**Reply by stephenafamo** [reply]
You can return an error from `BeginTx` to say that it does not support transactions. Since transactions are not used in SQLBoiler code, it should work well enough.
